ott/core/gw_barycenter.py

```diff
@@ -0,0 +1,323 @@
+    gw_convergence: Array of shape ``[max_iter,]`` containing the convergence
+      of all GW problems at each iteration.
+  """
+  x: Optional[jnp.ndarray] = None
+  c: Optional[jnp.ndarray] = None
+  a: Optional[jnp.ndarray] = None
+  errors: Optional[jnp.ndarray] = None
+  costs: Optional[jnp.ndarray] = None
+  gw_convergence: Optional[jnp.ndarray] = None
+
+  def set(self, **kwargs: Any) -> 'GWBarycenterState':
+    """Return a copy of self, possibly with overwrites."""
+    return self._replace(**kwargs)
+
+
+# TODO(michalk8): add citations
+@jax.tree_util.register_pytree_node_class
+class GromovWassersteinBarycenter(was_solver.WassersteinSolver):
+  """Gromov-Wasserstein barycenter solver.
+
+  Args:
+    epsilon: Entropy regulariser.
+    min_iterations: Minimum number of iterations.
+    max_iterations: Maximum number of outermost iterations.
+    threshold: Convergence threshold.
+    jit: Whether to jit the iteration loop.
+    store_inner_errors: Whether to store the errors of the GW solver, as well
+      as its linear solver, at each iteration for each measure.
+    quad_solver: The GW solver.
+    kwargs: Keyword argument for
+      :class:`ott.core.gromov_wasserstein.GromovWasserstein`.
+      Only used when ``quad_solver = None``.
+  """
+
+  def __init__(
+      self,
+      epsilon: Optional[float] = None,
+      min_iterations: int = 5,
+      max_iterations: int = 50,
+      threshold: float = 1e-3,
+      jit: bool = True,
+      store_inner_errors: bool = False,
+      quad_solver: Optional[gromov_wasserstein.GromovWasserstein] = None,
+      # TODO(michalk8): this maintains the API compatibility with `was_solver`
+      # but makes passing kwargs with the same name to `quad_solver` impossible
+      # will be fixed when refactoring the solvers
+      # note that `was_solver` also suffers from this
+      **kwargs: Any,
+  ):
+    super().__init__(
+        epsilon=epsilon,
+        min_iterations=min_iterations,
+        max_iterations=max_iterations,
+        threshold=threshold,
+        jit=jit,
+        store_inner_errors=store_inner_errors
+    )
+    self._quad_solver = quad_solver
+    if quad_solver is None:
+      kwargs["epsilon"] = epsilon
+      # TODO(michalk8): store only GW errors?
+      kwargs["store_inner_errors"] = store_inner_errors
+      self._quad_solver = gromov_wasserstein.GromovWasserstein(**kwargs)
+    assert not self._quad_solver.is_low_rank, "Low rank is not yet implemented."
+
+  def __call__(self, problem: GWBarycenterProblem, **kwargs: Any):
+    bar_fn = jax.jit(iterations, static_argnums=1) if self.jit else iterations
+    state = self.init_state(problem, **kwargs)
+    state = bar_fn(solver=self, problem=problem, init_state=state)
+    return self.output_from_state(state)
+
+  def init_state(
+      self,
+      problem: GWBarycenterProblem,
+      bar_init: Union[int, jnp.ndarray, Tuple[jnp.ndarray, jnp.ndarray]],
```

Comment:
this feels like it hides a bit more than just an init, but an important hyperparameter (size of the barycenter)

Reply:
maybe renaming this to keep the API you propose, or use something more like what is used for continuous_barycenter?